merlin-server/constraints-dsl-compiler/src/libs/constraints-ast.ts

```diff
@@ -1,38 +1,57 @@
 export enum NodeKind {
-  DummyConstraint = 'DummyConstraint',
-  ConstraintAnd = 'ConstraintAnd',
-  ConstraintOr = 'ConstraintOr'
+  ViolationsOf = 'ViolationsOf',
+  And = 'And',
+  Or = 'Or',
+  True = 'True'
 }
 
-export type Constraint = | DummyConstraint;
+/**
+ * Top-level constraints that directly produce violations
+ * when evaluated.
+ */
+export type Constraint = | ViolationsOf;
 
-export interface DummyConstraint {
-  kind: NodeKind.DummyConstraint,
-  someNumber: number,
+/**
+ * Generates violations whenever its expression generates a window.
+ */
+export interface ViolationsOf {
+  kind: NodeKind.ViolationsOf,
+  expression: WindowsExpression,
 }
 
-export type ConstraintSpecifier =
-    | Constraint
-    | ConstraintComposition
-    ;
-
 /**
- * Constraint Composition
+ * Operations that produce Windows when evaluated.
  *
- * Compose constraints together
+ * These are not valid top-level constraints, and
+ * typically need to be wrapped in `Constraint.ViolationsOf`.
+ */
+export type WindowsExpression =
+  | And
+  | Or
+  | True;
+
+/**
+ * Generates a violation when all of its sub-expressions generate windows.
  */
-export type ConstraintComposition =
-    | ConstraintAnd
-    | ConstraintOr
-    ;
+export interface And {
+  kind: NodeKind.And,
```

Comment:
I think we're going to have `And`/`Or` etc AST nodes on the Goal side of things too, so are the two different `And`/`Or` nodes going to be in the same trees such that we would want to further specify here with a kind like `"WindowExpressionAnd"`. I'm inclined to say "It doesn't hurt". I think contextual differences for nodes with the same tags become very confusing (see query vs aggregation contexts for mongodb's syntax).

Reply:
So are you recommending changing the type names too, or just the string in the `kind` field? Personally I'd find it pretty annoying to have to type `WindowsExpression.WindowsExpressionAnd(...)`, but I'm not sure that's what you're suggesting.

Reply:
Yes, just the strings. I think the eDSL should just maintain the shorthand `.and` and `.or`.

You can look at the scheduling eDSL for examples where the and/or functions are not static, but attached to the instances.

https://github.com/NASA-AMMOS/aerie/blob/develop/scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts#L22-L40

Reply:
done